optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/plugin/planner/SolverManager.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaweb.vehiclerouting.plugin.planner;
+
+import java.util.Arrays;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import org.optaplanner.core.api.solver.Solver;
+import org.optaplanner.core.api.solver.event.BestSolutionChangedEvent;
+import org.optaplanner.core.api.solver.event.SolverEventListener;
+import org.optaplanner.examples.vehiclerouting.domain.Vehicle;
+import org.optaplanner.examples.vehiclerouting.domain.VehicleRoutingSolution;
+import org.optaplanner.examples.vehiclerouting.domain.location.Location;
+import org.optaweb.vehiclerouting.plugin.planner.change.AddCustomer;
+import org.optaweb.vehiclerouting.plugin.planner.change.AddVehicle;
+import org.optaweb.vehiclerouting.plugin.planner.change.RemoveCustomer;
+import org.optaweb.vehiclerouting.plugin.planner.change.RemoveLocation;
+import org.optaweb.vehiclerouting.plugin.planner.change.RemoveVehicle;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.task.AsyncTaskExecutor;
+import org.springframework.stereotype.Component;
+
+/**
+ * Manages a solver running in a different thread.
+ * <p>
+ * Does following:
+ * <ul>
+ * <li>Starts solver by running {@link Solver#solve(Object problem)} in a thread that's not the caller's thread.</li>
+ * <li>Stops the solver (synchronously).</li>
+ * <li>Adds problem fact changes to the solver.</li>
+ * <li>Propagates any exception that happens in {@code Solver.solver()} (in a different thread) to the thread that
+ * interacts with {@code SolverManager}.</li>
+ * <li>Listens for best solution changes and publishes new best solutions via {@link SolutionPublisher}.</li>
+ * </ul>
+ */
+@Component
+class SolverManager implements SolverEventListener<VehicleRoutingSolution> {
```

Comment:
@MusaTalluzi FYI. This isn't meant to compete with your SolverManager (I just couldn't come up with a better name). TBH I haven't checked it out for a while so I'm not even sure whether and where they overlap. But definitely has a different goal (see the Javadoc). Please take a look and consider following questions:
- Does this bring something valuable that can be reused in your SolverManager API?
- Does your SM already provide this functionality? Can I drop this when SM is added to optaplanner-core?

Maybe this should be called AsyncSolver.

CC @ge0ffrey.

Reply:
Thanks Jiri for sharing this, it definitely gives us insights on what should be included in the API. In addition, you having to write this is another indication of the importance of providing SolverManager OOTB so it's less work for users in the future.

For your first question, stopping the solver is something that doesn't exist in the API, I'll add it. All the other features (except for ProblemFactChanges) already exists in SolverManager, I'll let you know before I add it to optaplanner-core to try it and suggest possible changes necessary for VRP.